### DISCUSSION TITLE
How do we containerize Goa apps properly?

### DISCUSSION THREAD

**Discussion by thecodekitchen** [goadesign/goa#3282] [Q&A] [answered]
I'm having trouble figuring out how to write a Dockerfile in order to containerize a Goa application. As far as I can tell, the entry point to launch the server is the extensionless file generated when I run 'go build' inside my 'cmd/<service-name>' directory. However, I'm unsure how to set up the port mapping for the Docker container. I'm also a bit confused about how I should package and distribute the CLI tool generated in the quickstart example. It would be nice if there was an example of this in the examples repo. Should I put in a feature request over there for that or am I over complicating things?

**Comment by gregwebs** [comment, marked as answer]
I don't think there is anything GOA specific here and there are many examples to go off of.

* GOA when used for HTTP produces an HTTP service like any other Go API/web framework.
* The CLI tool is a Go binary so it would be packaged like any other Go binary.

**Comment by gregwebs** [comment]
With that being said, generating OpenAPI is the core feature of GOA that most frameworks don't support well. If you want to serve these from your service you would need to add it to your Dockerfile, something like:

```go
# design/design.go
	Files("/openapi.json", "./gen/http/openapi.json")
	Files("/openapi3.json", "./gen/http/openapi3.json")
	Files("/swagger-ui/{*filepath}", "./assets/swagger-ui",
```

``` Dockerfile
# Dockerfile
ADD ./assets ./assets # for an OpenAPI GUI assets/swagger-ui
ADD ./gen/http/openapi.json ./gen/http/openapi.json
ADD ./gen/http/openapi3.json ./gen/http/openapi3.json
ADD ./gen/http/openapi.yaml ./gen/http/openapi.yaml
ADD ./gen/http/openapi3.yaml ./gen/http/openapi3.yaml
```
